# Tasks Notebook

### **Import libraries** 

In [173]:
import pandas as pd 
import numpy as np 

### **Task 1: Create Directory Structure**

In GitHub Codespaces using the terminal command line:

1. Navigate to the root of the repository: 
    ```bash
   cd /workspaces/computer-infrastructure-assessment
   
2. Create a `data` directory with two subdirectories: `timestamps` and `weather`:
    ```bash
   mkdir -p data/timestamps data/weather


### **Task 2: Timestamps**

1. Navigate to the `data/timestamps` directory:
   ```bash
    cd data/timestamps
   
2. Use the `date` command to output the current date and time to a text file named `now.txt`:
   ```bash
   date >> now.txt
   
3. Repeat this 10 times by creating a for loop that appends to the file rather than overwriting it:
   ```bash
   for i in {1..10}; do date >> now.txt; done
   
4. Use the `cat` command to read and verify the contents of `now.txt`:
   ```bash
   cat now.txt

[Reference: 'Date' command in Linux](https://www.geeksforgeeks.org/date-command-linux-examples/t)<br/>
[Reference: Write current date/time to a file](https://stackoverflow.com/questions/43221469/write-current-date-time-to-a-file-using-shell-script)<br/>
[Reference: Run date command a number of times in Linux](https://www.cyberciti.biz/faq/bsd-appleosx-linux-bash-shell-run-command-n-times/)

### **Task 3: Formatting Timestamps**

1. From Task 2, we are still in the `data/timestamps` directory.

   
2. Use the `date` command with the desired format to print the current date and time:
   ```bash
   date +"%Y%m%d_%H%M%S"`
   
3. Use the command from step 2 with the `>>` operator to append the output into a file named `formatted.txt`:
   ```bash
   date +"%Y%m%d_%H%M%S" >> formatted.txt`
   
4. Use the `cat` command to read and verify the contents of `formatted.txt`:
   ```bash
   cat formatted.txt`

[Reference: Date formatting in Linux](https://phoenixnap.com/kb/linux-date-command)


### **Task 4: Create Timestamped Files**

1. Ensure you are in the `data/timestamps` directory from Task 3.

2. Use the `touch` command combined with the `date` command to create an empty file named with the current date and timestamp in the format `YYYYmmdd_HHMMSS`:
   ```bash
   touch $(date +%Y%m%d_%H%M%S).txt

[Reference: How to create a file with today's date in the filename](https://stackoverflow.com/questions/48270960/how-to-create-a-file-with-todays-date-in-the-filename) <br/>
[Reference: How to create a file with today's date in the filename (2)](https://superuser.com/questions/399475/in-linux-how-to-make-a-file-with-a-name-that-is-current-date-and-time)<br/>

### **Task 5: Download Today's Weather Data**

1. Navigate to the `weather` directory:
    ```bash
   cd data/weather
   
2. Use the `wget` command to download the file from the web. Use the `-O` option to name the file where the data will be saved:
   ```bash
   wget -O weather.json https://prodapi.metweb.ie/observations/athenry/today

3. Verify that the file has been created:
   ```bash
    ls

[Reference: Use wget to download and name file](https://stackoverflow.com/questions/16678487/how-to-change-filename-of-a-file-downloaded-with-wget)

### **Task 6: Timestamp the Data**

1. From Task 5, we are still in the `data/weather` directory.
   
2. Modify the command from Task 5 to name the JSON file with today’s date and timestamp:
   ```bash
   wget -O $(date +%Y%m%d_%H%M%S).json https://prodapi.metweb.ie/observations/athenry/today

3. Verify that the file has been created:
   ```bash
   ls


### **Task 7: Write the Script**

1. Navigate to the root of the repository:
   ```bash
     cd /workspaces/computer-infrastructure-assessment
     
2. Use `nano` to create the script:
   ```bash
     nano weather.sh
     
3. In the `nano` editor, write the following code to create the script:

   ```bash
     #!/bin/bash
     
     [Indicates that the script should be run in the Bash shell]

   
     wget -O data/weather/`date +%Y%m%d_%H%M%S`.json https://prodapi.metweb.ie/observations/athenry/today
     
     [Changes to the `data/weather` directory then downloads and saves the weather data there in a timestamped file]

   
     echo "Downloaded weather data:"
     
     [Prints a message to confirm the file has been downloaded]

4. Save and exit `nano` then use the `chmod` command to make the script executable to user:
   ```bash
     chmod u+x weather.sh
     
5. Run the script:
   ```bash
     ./weather.sh
     
6. Navigate to the `data/weather` directory to check that the script has created the file:
   ```bash
     cd data/weather
     
[Reference: Create shell script in nano](https://www.linuxandubuntu.com/home/how-to-create-shell-scripts/)<br/>
[Reference: How to Make Script Executable in Linux | chmod Command](https://www.linuxandubuntu.com/home/how-to-create-shell-scripts/)<br/>
[Reference: How to execute a shell script in Linux](https://www.cyberciti.biz/faq/how-to-execute-a-shell-script-in-linux/)<br/>


### **Task 9: Data Analysis with Pandas**

In [174]:
# Load the JSON file into a DataFrame
df = pd.read_json('data/weather/20241031_213107.json')

# Preview the first few rows of data
df.head()


name  temperature symbol weatherDescription    text  windSpeed windGust  \
0  Athenry           11    15n         Fog / Mist  "Mist"          9        -   
1  Athenry           11    15n         Fog / Mist  "Mist"          9        -   
2  Athenry           11    15n         Fog / Mist  "Mist"          9        -   
3  Athenry           10    15n         Fog / Mist  "Mist"         11        -   
4  Athenry           10    15n         Fog / Mist  "Mist"         13        -   

  cardinalWindDirection  windDirection  humidity  rainfall  pressure  \
0                    SW            225        97       0.0      1028   
1                     S            180        98       0.1      1028   
2                     S            180        97       0.0      1028   
3                     S            180        97       0.0      1027   
4                    SW            225        98       0.1      1027   

    dayName       date reportTime  
0  Thursday 2024-10-31      00:00  
1  Thursday 2024-10-31      01:00  
2  Thursday 2024-10-31      02:00  
3  Thursday 2024-10-31      03:00  
4  Thursday 2024-10-31      04:00

[Reference:How to Read JSON Files with Pandas?](https://www.geeksforgeeks.org/how-to-read-json-files-with-pandas/)<br/>

In [175]:
total_rows = len(df)
print(f"Total number of rows: {total_rows}")

Total number of rows: 22


In [176]:
# Opt into future behavior for downcasting as error was recieved for downcasting behaviour for 'replace' being deprecated in future.
pd.set_option('future.no_silent_downcasting', True)

# Replace blank, whitespace-only fields, and '-' with np.nan
df.replace(r'^\s*$', np.nan, regex=True, inplace=True)  # Handle blank or whitespace-only
df.replace('-', np.nan, inplace=True)  # Replace '-' specifically

# Count total rows and missing values by column
total_rows = len(df)
missing_data = df.isnull().sum()

# Display the results
print(f"Total number of rows: {total_rows}")
print("Columns with missing values:")
print(missing_data[missing_data > 0])


Total number of rows: 22
Columns with missing values:
windGust    22
dtype: int64


[Reference: Replace values in Pandas dataframe using regex](https://www.geeksforgeeks.org/replace-values-in-pandas-dataframe-using-regex/)<br/>
[Reference: Pandas: How to Replace Zero with NaN](https://www.statology.org/pandas-replace-0-with-nan/)<br/>
[Reference: A Guide to R Regular Expressions](https://www.datacamp.com/tutorial/regex-r-regular-expressions-guide) <br/>
[Reference: Using isnull() and groupby() on a pandas dataframe](https://stackoverflow.com/questions/46106954/using-isnull-and-groupby-on-a-pandas-dataframe)<br/>
[Reference: Display rows with one or more NaN values in pandas dataframe](https://stackoverflow.com/questions/43424199/display-rows-with-one-or-more-nan-values-in-pandas-dataframe)<br/>
[Reference: pd.NA vs np.nan for pandas](https://stackoverflow.com/questions/60115806/pd-na-vs-np-nan-for-pandas)<br/>
[Reference: Replace values in Pandas dataframe using regex](https://www.geeksforgeeks.org/replace-values-in-pandas-dataframe-using-regex/)<br/>
[Reference: How to solve pandas (2.2.0) "FutureWarning: Downcasting behavior in replace is deprecated".](https://github.com/pandas-dev/pandas/issues/57734)<br/>

In [177]:
# Get a summary of the DataFrame, including column data types and counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   name                   22 non-null     object        
 1   temperature            22 non-null     int64         
 2   symbol                 22 non-null     object        
 3   weatherDescription     22 non-null     object        
 4   text                   22 non-null     object        
 5   windSpeed              22 non-null     int64         
 6   windGust               0 non-null      object        
 7   cardinalWindDirection  22 non-null     object        
 8   windDirection          22 non-null     int64         
 9   humidity               22 non-null     int64         
 10  rainfall               22 non-null     float64       
 11  pressure               22 non-null     int64         
 12  dayName                22 non-null     object        
 13  date   

In [178]:
# Generate summary statistics for numerical columns
df.describe()

temperature  windSpeed  windDirection   humidity   rainfall  \
count    22.000000  22.000000      22.000000  22.000000  22.000000   
mean     10.545455  10.090909     214.772727  97.045455   0.063636   
min      10.000000   7.000000     180.000000  94.000000   0.000000   
25%      10.000000   9.000000     225.000000  97.000000   0.000000   
50%      10.000000  11.000000     225.000000  97.000000   0.100000   
75%      11.000000  11.000000     225.000000  98.000000   0.100000   
max      12.000000  13.000000     225.000000  99.000000   0.200000   
std       0.670982   1.715728      19.301941   1.252703   0.065795   

          pressure                 date  
count    22.000000                   22  
mean   1026.409091  2024-10-31 00:00:00  
min    1024.000000  2024-10-31 00:00:00  
25%    1025.250000  2024-10-31 00:00:00  
50%    1027.000000  2024-10-31 00:00:00  
75%    1027.000000  2024-10-31 00:00:00  
max    1028.000000  2024-10-31 00:00:00  
std       1.140555                  NaN

In [179]:
# Frequency of each weather description
description_counts = df['weatherDescription'].value_counts()
print(description_counts)

weatherDescription
Fog / Mist      15
Cloudy           3
Rain showers     2
Light rain       2
Name: count, dtype: int64


[Reference:Pandas Index.value_counts()](https://www.geeksforgeeks.org/python-pandas-index-value_counts/)<br/>

In [180]:
# Examine weather descriptions over time
df[['date', 'reportTime', 'weatherDescription']].set_index('date')


reportTime weatherDescription
date                                    
2024-10-31      00:00         Fog / Mist
2024-10-31      01:00         Fog / Mist
2024-10-31      02:00         Fog / Mist
2024-10-31      03:00         Fog / Mist
2024-10-31      04:00         Fog / Mist
2024-10-31      05:00         Fog / Mist
2024-10-31      06:00         Fog / Mist
2024-10-31      07:00       Rain showers
2024-10-31      08:00         Fog / Mist
2024-10-31      09:00       Rain showers
2024-10-31      10:00         Fog / Mist
2024-10-31      11:00         Fog / Mist
2024-10-31      12:00         Fog / Mist
2024-10-31      13:00         Fog / Mist
2024-10-31      14:00         Fog / Mist
2024-10-31      15:00        Light rain 
2024-10-31      16:00        Light rain 
2024-10-31      17:00         Fog / Mist
2024-10-31      18:00         Fog / Mist
2024-10-31      19:00             Cloudy
2024-10-31      20:00             Cloudy
2024-10-31      21:00             Cloudy

[Reference:Pandas DataFrame set_index() Method](https://www.w3schools.com/python/pandas/ref_df_set_index.asp)<br/>

In [181]:
# Frequency of each cardinal wind direction
cardinal_direction_counts = df['cardinalWindDirection'].value_counts()
print(cardinal_direction_counts)


cardinalWindDirection
SW    17
S      5
Name: count, dtype: int64


[Reference:Pandas DataFrame set_index() Method](https://www.w3schools.com/python/pandas/ref_df_set_index.asp)<br/>

In [182]:
# Examine weather descriptions over time
df[['date', 'reportTime', 'cardinalWindDirection']].set_index('date')

reportTime cardinalWindDirection
date                                       
2024-10-31      00:00                    SW
2024-10-31      01:00                     S
2024-10-31      02:00                     S
2024-10-31      03:00                     S
2024-10-31      04:00                    SW
2024-10-31      05:00                    SW
2024-10-31      06:00                    SW
2024-10-31      07:00                    SW
2024-10-31      08:00                    SW
2024-10-31      09:00                    SW
2024-10-31      10:00                    SW
2024-10-31      11:00                    SW
2024-10-31      12:00                    SW
2024-10-31      13:00                    SW
2024-10-31      14:00                    SW
2024-10-31      15:00                    SW
2024-10-31      16:00                    SW
2024-10-31      17:00                    SW
2024-10-31      18:00                     S
2024-10-31      19:00                    SW
2024-10-31      20:00                    SW
2024-10-31      21:00                     S

[Reference:Pandas DataFrame set_index() Method](https://www.w3schools.com/python/pandas/ref_df_set_index.asp)<br/>

## Summary and Insights

### Data Structure
- The dataset contains **22 rows** and **15 columns**.
- Each row represents a single weather observation at a specific timestamp.
### Data Completeness
- No missing values in any columns.
### Key Columns
- **Name**: Station name (Athenry).
- **Dry Bulb Temperature**: Temperature in whole degrees Celsius.
- **Weather Description**: Text description of the weather conditions.
- **Wind Speed**: Wind speed in knots (kt).
- **Cardinal Wind Direction**: Wind direction in cardinal form.
- **Relative Humidity**: Humidity as a percentage.
- **Rainfall**: Precipitation amount in millimeters (mm).
- **Pressure**: Mean sea level pressure in millibars (mbar).
- **Day Name**: The day of the week for the observation.
- **Date**: Date of the observation.
- **Time of Observation**: Local time when the observation was recorded.

[Reference: Description of metrics deom Met Eireann](https://data.gov.ie/dataset/todays-weather-athenry/resource/fffca98d-f3b6-47de-a834-837f0633e14b)

## Daily Weather Summary for 2024-10-31 at Athenry Station

- **Temperature**: Cool and steady, averaging 10.5°C, with a range from 10°C to 12°C.
- **Weather Description**: Mostly "Fog / Mist" throughout the day, with some periods of light rain and showers.
- **Wind**: Gentle, averaging 10 km/h, primarily from the south to southwest.
- **Humidity**: Very high, averaging 97%, consistent with the frequent "fog and mist" conditions.
- **Rainfall**: Minimal, with a maximum of 0.2 mm and an average of 0.06 mm.

### Overall
A cool, humid day with consistent south-southwest winds, frequent fog or mist, and very light, intermittent precipitation.


## Potential Future Analysis
- Trend analysis over the day, such as changes in temperature or humidity.
- Correlation analysis, such as between wind speed and temperature changes.